packages/cli/src/modules/mcp/tools/execute-workflow.tool.ts

```diff
@@ -0,0 +1,342 @@
+			const isTimeout = error instanceof UserError && error.message.includes('timed out');
+			const output: ExecuteWorkflowOutput = {
+				success: false,
+				executionId: null,
+				error: isTimeout
+					? `Workflow execution timed out after ${WORKFLOW_EXECUTION_TIMEOUT_MS / 1000} seconds`
+					: error.message,
+			};
+
+			telemetryPayload.results = {
+				success: false,
+				error: isTimeout ? 'Workflow execution timed out' : error.message,
+			};
+			telemetry.track(USER_CALLED_MCP_TOOL_EVENT, telemetryPayload);
+
+			return {
+				content: [{ type: 'text', text: jsonStringify(output) }],
+				structuredContent: output,
+			};
+		}
+	},
+});
+
+/**
+ * Executes a workflow for the given user with provided inputs.
+ * In order to "synchronously" execute the workflow,
+ * it is mapping mcp tool inputs to trigger node pin data and starting execution from there.
+ * LIMITATION: Does not properly support workflows with multiple triggers.
+ */
+export const executeWorkflow = async (
+	user: User,
+	workflowFinderService: WorkflowFinderService,
+	activeExecutions: ActiveExecutions,
+	workflowRunner: WorkflowRunner,
+	workflowId: string,
+	inputs?: z.infer<typeof inputSchema>['inputs'],
+): Promise<ExecuteWorkflowOutput> => {
+	const workflow = await workflowFinderService.findWorkflowForUser(workflowId, user, [
+		'workflow:execute',
+	]);
+
+	if (!workflow || workflow.isArchived) {
+		throw new UserError('Workflow not found');
+	}
+
+	if (!workflow.settings?.availableInMCP) {
+		throw new UserError(
+			'Workflow is not available for execution via MCP. Enable access in the workflow settings to make it available.',
+		);
+	}
+
+	const triggerNode = findMcpSupportedTrigger(workflow);
+
+	if (!triggerNode) {
+		throw new UserError(
+			`Only workflows with the following trigger nodes can be executed: ${Object.values(SUPPORTED_MCP_TRIGGERS).join(', ')}.`,
+		);
+	}
+
+	const runData: IWorkflowExecutionDataProcess = {
+		executionMode: getExecutionModeForTrigger(triggerNode),
+		workflowData: workflow,
+		userId: user.id,
+	};
+
+	// Set the trigger node as the start node and pin data for it
+	// This will enable us to run the workflow from the trigger node with the provided inputs without waiting for an actual trigger event
+	runData.startNodes = [{ name: triggerNode.name, sourceData: null }];
+	runData.pinData = getPinDataForTrigger(triggerNode, inputs);
+
+	const executionId = await workflowRunner.run(runData);
+
+	// Create a timeout promise
+	let timeoutId: NodeJS.Timeout | undefined;
+	const timeoutPromise = new Promise<never>((_, reject) => {
+		timeoutId = setTimeout(() => {
+			reject(
+				new UserError(
+					`Workflow execution timed out after ${WORKFLOW_EXECUTION_TIMEOUT_MS / 1000} seconds`,
+				),
+			);
+		}, WORKFLOW_EXECUTION_TIMEOUT_MS);
+	});
+
+	try {
+		// Race between the execution completion and the timeout
+		const data = await Promise.race([
+			activeExecutions.getPostExecutePromise(executionId),
+			timeoutPromise,
+		]);
```

Comment:
Would it be possible to reuse existing timeout functionality? 🤔 

<img width="1676" height="186" alt="Capture 2025-11-14 at 16 22 52@2x" src="https://github.com/user-attachments/assets/0a81864c-a0f6-42e5-9845-4f4b7f43b6ce" />